OpenRA.Mods.Common/Orders/ChromeLogicOrderGenerator.cs

```diff
@@ -0,0 +1,43 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.Orders
+{
+	public abstract class ChromeLogicOrderGenerator : IOrderGenerator
```

Comment:
What is the point of this class? Its duplicate of `OrderGenerator` the only change is that `Order()` is made virtual which is overridden in `RepairOrderGenerator `. So there is no logic in this class needed in `RepairOrderGenerator `. IMO it would be better to let `RepairOrderGenerator ` implement `IOrderGenerator` directly.

Reply:
Becuase `PowerDownOrderGenerator` or even `SellOrderGenerator` can also inherit this class for mouse dragging if somebody want later. 

Although it seems kinda senseless for now, but I hope `ChromeLogicOrderGenerator` can replace `GlobalButtonOrderGenerator<T>` that `PowerDownOrderGenerator` and `SellOrderGenerator` inherit.  `GlobalButtonOrderGenerator<T>` inheirts `OrderGenerator`, which blocks some important mouse input event for mouse dragging like `button up`, as well as cannot change due to not `virtual`. In addition, `GlobalButtonOrderGenerator<T>` is not suitable on achieving different operations based on functions in detail, like slight diffences in mechanic when use `OrderInner()`, at present.

If both powerdown and even selling use dragging in the future, I think devs have to make some big changes to `GlobalButtonOrderGenerator<T>`, or they can just use `ChromeLogicOrderGenerator`. **In conclusion, for those global buttons (repair/sell/powerdown) have similary design on implement, I think there should be a parent class for them to inherit** but also allow more diffences in comprasion of `GlobalButtonOrderGenerator<T>` at present.

That's why I decide to build a new class for `RepairOrderGenerator` to inherit, or probably later two other classes to inherit. If `GlobalButtonOrderGenerator<T>` become more open and useful, I will also abandon `ChromeLogicOrderGenerator ` and put `RepairOrderGenerator` into this class family.

Reply:
I think I should change this class name( I have to admit that there is no 'logic' in this class) and some of the function, and make a mouse dragging power down mode to persuade you.

Reply:
Then it would make more sense to move the code up from `RepairOrderGenerator.Order()` to this class, right?

Reply:
I gonna to do:

1. Change the name of `ChromeLogicOrderGenerator` into `GlobalButtonOrderGenerator`, delete orignal `GlobalButtonOrderGenerator<T>`.

2. Rewrite other two global buttons' OG in order to inherit new `GlobalButtonOrderGenerator`, as well as new drag function for Powerdown mode.

3. Not going to move the code up from `RepairOrderGenerator.Order()`. Instead, the `SellOrderGenerator.Order()`. I will still make the parent function `virtual` due to most of global buttons born with no dragging function like most of click-then-release buttons in orignal C&C. If later somebody going to make new global buttons, they will probably start with no dragging function and using  `Order()` directly from parent class.